File: scripts/generate_paper_ref05.R
suppressPackageStartupMessages({
  library(edgeR)
  library(jsonlite)
  library(limma)
  library(speckle)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 1) stop("usage: generate_paper_ref05.R CONFIG.json")
cfg <- fromJSON(args[[1]], simplifyVector = TRUE)
dir.create(cfg$output_dir, recursive = TRUE, showWarnings = FALSE)

read_counts <- function(path) {
  table <- read.delim(path, check.names = FALSE, stringsAsFactors = FALSE)
  genes <- as.character(table[[1]])
  if (anyDuplicated(genes)) stop("pseudobulk gene identities are duplicated")
  matrix <- as.matrix(table[, -1, drop = FALSE])
  storage.mode(matrix) <- "integer"
  rownames(matrix) <- genes
  if (any(!is.finite(matrix)) || any(matrix < 0)) stop("invalid pseudobulk counts")
  matrix
}

read_samples <- function(path, counts) {
  samples <- read.delim(path, check.names = FALSE, stringsAsFactors = FALSE)
  required <- c("sample_id", "donor", "condition", "n_cells")
  if (!all(required %in% colnames(samples))) stop("sample metadata is incomplete")
  if (!setequal(samples$sample_id, colnames(counts))) stop("sample/count mismatch")
  samples <- samples[match(colnames(counts), samples$sample_id), , drop = FALSE]
  if (anyDuplicated(samples$sample_id)) stop("sample identities are duplicated")
  samples
}

paired_design <- function(samples, condition_column = "condition") {
  samples$donor <- factor(samples$donor)
  samples[[condition_column]] <- factor(
    samples[[condition_column]], levels = c(cfg$baseline, cfg$target)
  )
  if (any(is.na(samples[[condition_column]]))) stop("unknown condition label")
  tab <- table(samples$donor, samples[[condition_column]])
  if (any(tab != 1L)) stop("each donor must have exactly one sample per condition")
  design <- model.matrix(
    reformulate(c("donor", condition_column)),
    data = samples
  )
  if (qr(design)$rank < ncol(design)) stop("paired design is rank deficient")
  coefficient <- grep(paste0("^", condition_column), colnames(design))
  if (length(coefficient) != 1) stop("condition coefficient is not unique")
  list(samples = samples, design = design, coefficient = coefficient)
}

paired_propeller_design <- function(samples, condition_column = "condition") {
  samples$donor <- factor(samples$donor)
  samples[[condition_column]] <- factor(
    samples[[condition_column]], levels = c(cfg$baseline, cfg$target)
  )
  if (any(is.na(samples[[condition_column]]))) stop("unknown condition label")
  tab <- table(samples$donor, samples[[condition_column]])
  if (any(tab != 1L)) stop("each donor must have exactly one sample per condition")

  # propeller.ttest requires a no-intercept design with the two group-specific
  # columns first. Additional donor columns are accepted as confounders.
  design <- model.matrix(
    reformulate(c(condition_column, "donor"), intercept = FALSE),
    data = samples
  )
  if (qr(design)$rank < ncol(design)) stop("paired Propeller design is rank deficient")
  group_columns <- match(
    paste0(condition_column, c(cfg$baseline, cfg$target)),
    colnames(design)
  )
  if (any(is.na(group_columns))) stop("Propeller group columns are not uniquely estimable")

  # Speckle expects a vector whose length equals ncol(design), not a one-column
  # limma contrast matrix. The remaining zeroes retain donor adjustment.
  contrast <- numeric(ncol(design))
  names(contrast) <- colnames(design)
  contrast[group_columns[[1]]] <- -1
  contrast[group_columns[[2]]] <- 1
  list(samples = samples, design = design, contrast = contrast)
}

run_edger <- function(counts, samples, condition_column = "condition") {
  designed <- paired_design(samples, condition_column)
  y <- DGEList(counts = counts, samples = designed$samples)
  keep <- filterByExpr(y, design = designed$design)
  if (!any(keep)) stop("filterByExpr retained no genes")
  y <- y[keep, , keep.lib.sizes = FALSE]
  y <- calcNormFactors(y)
  y <- estimateDisp(y, designed$design)
  fit <- glmQLFit(y, designed$design)
  test <- glmQLFTest(fit, coef = designed$coefficient)
  result <- topTags(test, n = Inf, sort.by = "none")$table
  result$gene <- rownames(result)
  result <- result[, c("gene", "logFC", "F", "PValue", "FDR")]
  list(result = result, design = designed$design, samples = designed$samples)
}

write_tsv <- function(value, path) {
  write.table(value, path, sep = "\t", row.names = FALSE, quote = FALSE, na = "")
}

evaluation_counts <- read_counts(cfg$evaluation_counts)
evaluation_samples <- read_samples(cfg$evaluation_samples, evaluation_counts)
cat("REF-05-A: fitting donor-aware evaluation edgeR\n")
flush.console()
evaluation_fit <- run_edger(evaluation_counts, evaluation_samples)
write_tsv(evaluation_fit$result, file.path(cfg$output_dir, "edger_reference.tsv"))
edge_design <- data.frame(
  sample_id = evaluation_fit$samples$sample_id,
  evaluation_fit$design,
  check.names = FALSE
)
write_tsv(edge_design, file.path(cfg$output_dir, "edger_design_matrix.tsv"))
writeLines(capture.output(sessionInfo()), file.path(cfg$output_dir, "edger_session_info.txt"))

cells <- read.delim(cfg$evaluation_cells, check.names = FALSE, stringsAsFactors = FALSE)
required_cells <- c("sample_id", "donor", "condition", "state")
if (!all(required_cells %in% colnames(cells))) stop("cell metadata is incomplete")
sample_table <- unique(cells[, c("sample_id", "donor", "condition")])
if (anyDuplicated(sample_table$sample_id)) stop("sample IDs map to multiple designs")
cat("REF-05-B: fitting donor-aware evaluation Propeller\n")
flush.console()
prop_list <- getTransformedProps(
  clusters = as.character(cells$state),
  sample = as.character(cells$sample_id)
)
sample_order <- colnames(prop_list$Proportions)
sample_table <- sample_table[match(sample_order, sample_table$sample_id), , drop = FALSE]
if (any(is.na(sample_table$sample_id))) stop("Propeller sample alignment failed")
designed <- paired_propeller_design(sample_table)
rownames(designed$design) <- sample_table$sample_id
propeller <- propeller.ttest(
  prop_list,
  design = designed$design,
  contrasts = designed$contrast,
  robust = TRUE,
  trend = FALSE,
  sort = FALSE
)
if (!"cluster" %in% colnames(propeller)) propeller$cluster <- rownames(propeller)
propeller <- propeller[match(rownames(prop_list$Proportions), propeller$cluster), , drop = FALSE]
baseline_samples <- sample_table$condition == cfg$baseline
target_samples <- sample_table$condition == cfg$target
baseline_mean <- rowMeans(prop_list$Proportions[, baseline_samples, drop = FALSE])
target_mean <- rowMeans(prop_list$Proportions[, target_samples, drop = FALSE])
p_column <- intersect(c("P.Value", "PValue", "p.value", "pvalue"), colnames(propeller))
if (length(p_column) == 0) stop("Propeller result lacks a p-value column")
if (!"FDR" %in% colnames(propeller)) stop("Propeller result lacks FDR")
propeller_reference <- data.frame(
  cluster = as.character(propeller$cluster),
  baseline_proportion = as.numeric(baseline_mean[propeller$cluster]),
  target_proportion = as.numeric(target_mean[propeller$cluster]),
  effect = as.numeric(target_mean[propeller$cluster] - baseline_mean[propeller$cluster]),
  PValue = as.numeric(propeller[[p_column[[1]]]]),
  FDR = as.numeric(propeller$FDR),
  stringsAsFactors = FALSE
)
write_tsv(propeller_reference, file.path(cfg$output_dir, "propeller_reference.tsv"))
propeller_design <- data.frame(
  sample_id = sample_table$sample_id,
  designed$design,
  check.names = FALSE
)
write_tsv(propeller_design, file.path(cfg$output_dir, "propeller_design_matrix.tsv"))
writeLines(capture.output(sessionInfo()), file.path(cfg$output_dir, "propeller_session_info.txt"))

calibration_counts <- read_counts(cfg$calibration_counts)
calibration_samples <- read_samples(cfg$calibration_samples, calibration_counts)
donors <- sort(unique(as.character(calibration_samples$donor)))
if (length(donors) < 3L) stop("null calibration requires at least three donors")
masks <- 1:(2^length(donors) - 2L)
method_null <- list()
replicate_null <- list()
cat("REF-05-C: fitting ", length(masks), " mixed donor-pair null permutations\n", sep = "")
flush.console()
for (mask in masks) {
  permuted <- calibration_samples
  swapped <- donors[as.logical(intToBits(mask)[seq_along(donors)])]
  permuted$condition_perm <- as.character(permuted$condition)
  for (donor in swapped) {
    member <- permuted$donor == donor
    permuted$condition_perm[member] <- ifelse(
      permuted$condition_perm[member] == cfg$baseline,
      cfg$target,
      cfg$baseline
    )
  }
  fitted <- run_edger(calibration_counts, permuted, "condition_perm")
  table <- fitted$result
  table$permutation_id <- sprintf("swap_%0*d", length(donors), mask)
  table$swapped_donors <- paste(swapped, collapse = ",")
  method_null[[length(method_null) + 1L]] <- table[, c(
    "permutation_id", "swapped_donors", "gene", "logFC", "F", "PValue", "FDR"
  )]
  replicate_null[[length(replicate_null) + 1L]] <- data.frame(
    permutation_id = sprintf("swap_%0*d", length(donors), mask),
    swapped_donors = paste(swapped, collapse = ","),
    n_tested_genes = nrow(table),
    type1_rate = mean(table$PValue <= 0.05),
    null_effect_bias = median(table$logFC),
    median_absolute_effect = median(abs(table$logFC)),
    exchangeability_violation_count = 0L,
    cell_label_permutation = FALSE,
    stringsAsFactors = FALSE
  )
  cat("REF-05-C: completed donor-swap permutation ", mask, "/", max(masks), "\n", sep = "")
  flush.console()
}
method_null <- do.call(rbind, method_null)
replicate_null <- do.call(rbind, replicate_null)
write_tsv(method_null, file.path(cfg$output_dir, "method_null_reference.tsv"))
write_tsv(replicate_null, file.path(cfg$output_dir, "replicate_null_reference.tsv"))

write_json(
  list(
    R = paste(R.version$major, R.version$minor, sep = "."),
    edgeR = as.character(packageVersion("edgeR")),
    limma = as.character(packageVersion("limma")),
    speckle = as.character(packageVersion("speckle")),
    jsonlite = as.character(packageVersion("jsonlite"))
  ),
  file.path(cfg$output_dir, "r_environment.json"),
  auto_unbox = TRUE,
  pretty = TRUE
)

cat("REF-05-A: edgeR genes=", nrow(evaluation_fit$result), "\n", sep = "")
cat("REF-05-B: Propeller states=", nrow(propeller_reference), "\n", sep = "")
cat("REF-05-C: donor-swap permutations=", length(masks), "\n", sep = "")

File: scripts/generate_paper_ref05.py
from __future__ import annotations

import argparse
import csv
import gzip
import hashlib
import json
import math
import subprocess
from collections import Counter
from pathlib import Path
from typing import Any, Iterable


SCHEMA_VERSION = "pertura-paper-ref05-v1"
DATASET_ID = "kang18_8vs8_pbmc"
SEED = 1729
BASELINE = "ctrl"
TARGET = "stim"


def _sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for block in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(block)
    return "sha256:" + digest.hexdigest()


def _write_json(path: Path, payload: Any) -> None:
    path.write_text(
        json.dumps(payload, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
        newline="\n",
    )


def _write_tsv(path: Path, fieldnames: list[str], rows: Iterable[dict[str, Any]]) -> None:
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(
            handle,
            fieldnames=fieldnames,
            delimiter="\t",
            lineterminator="\n",
            extrasaction="ignore",
        )
        writer.writeheader()
        writer.writerows(rows)


def _read_tsv(path: Path) -> tuple[list[str], list[dict[str, str]]]:
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle, delimiter="\t")
        return list(reader.fieldnames or []), [dict(row) for row in reader]


def _selection_rows(splits_path: Path, split: str) -> list[dict[str, str]]:
    payload = json.loads(splits_path.read_text(encoding="utf-8"))
    record = payload["datasets"][DATASET_ID][split]
    path = splits_path.resolve().parent.parent / record["cell_selection_path"]
    if not path.is_file() or _sha256(path) != record["cell_selection_file_sha256"]:
        raise ValueError(f"Kang {split} selection hash drift")
    with gzip.open(path, "rt", encoding="utf-8", newline="") as handle:
        rows = [dict(row) for row in csv.DictReader(handle, delimiter="\t")]
    required = {"cell_id", "unit_id", "is_control"}
    if not rows or not required.issubset(rows[0]):
        raise ValueError(f"Kang {split} selection is missing required fields")
    identities = [row["cell_id"] for row in rows]
    if len(identities) != len(set(identities)):
        raise ValueError(f"duplicate Kang {split} cell identity")
    return rows


def _split_design(rows: list[dict[str, str]], split: str) -> dict[str, str]:
    design: dict[str, str] = {}
    arms: dict[str, set[str]] = {}
    for row in rows:
        cell = str(row["cell_id"])
        donor = str(row["unit_id"])
        control_flag = str(row["is_control"]).lower()
        if control_flag not in {"true", "false"}:
            raise ValueError(f"Kang {split} has an invalid control flag for {cell}")
        condition = BASELINE if control_flag == "true" else TARGET
        if not donor:
            raise ValueError(f"Kang {split} contains an empty donor identity")
        design[cell] = donor + "\x1f" + condition
        arms.setdefault(donor, set()).add(condition)
    incomplete = sorted(donor for donor, values in arms.items() if values != {BASELINE, TARGET})
    if incomplete:
        raise ValueError(f"Kang {split} donors lack paired conditions: {incomplete}")
    if len(arms) < 3:
        raise ValueError(f"Kang {split} requires at least three paired donors")
    return design


def _condition_matches(value: str, condition: str) -> bool:
    normalized = value.strip().lower()
    if condition == BASELINE:
        return normalized in {"ctrl", "control", "unstim", "unstimulated"}
    return normalized in {"stim", "stimulated"}


def _materialize_split_inputs(
    source: Any,
    selection: list[dict[str, str]],
    output_dir: Path,
    split: str,
    *,
    chunk_rows: int = 1024,
) -> dict[str, Any]:
    import numpy as np
    from scipy import sparse

    required_obs = {"ind", "stim", "cell"}
    missing_obs = sorted(required_obs - set(source.obs.columns))
    if missing_obs:
        raise ValueError("Kang H5AD is missing: " + ", ".join(missing_obs))
    split_design = _split_design(selection, split)
    source_names = source.obs_names.astype(str)
    requested = set(split_design)
    missing_cells = sorted(requested.difference(source_names))
    if missing_cells:
        raise ValueError(f"Kang artifact is missing {len(missing_cells)} {split} cells")
    selected = np.flatnonzero(source_names.isin(requested))
    selected_cells = [str(source_names[index]) for index in selected]
    sample_keys = sorted(set(split_design[cell] for cell in selected_cells))
    sample_index = {key: index for index, key in enumerate(sample_keys)}
    counts = np.zeros((len(sample_keys), int(source.n_vars)), dtype=np.int64)
    sample_cell_counts: Counter[str] = Counter()
    cell_rows: list[dict[str, str]] = []

    print(
        f"REF-05: sequentially scanning {source.n_obs} Kang rows for "
        f"{len(selected)} {split} cells",
        flush=True,
    )
    for start in range(0, int(source.n_obs), chunk_rows):
        stop = min(start + chunk_rows, int(source.n_obs))
        in_chunk = selected[(selected >= start) & (selected < stop)]
        if not len(in_chunk):
            continue
        block = source.X[start:stop, :]
        if hasattr(block, "to_memory"):
            block = block.to_memory()
        block = block[in_chunk - start, :]
        block = block.tocsr() if sparse.issparse(block) else np.asarray(block)
        values = block.data if sparse.issparse(block) else block.reshape(-1)
        if values.size and (
            not np.isfinite(values).all()
            or (values < 0).any()
            or not np.allclose(values, np.rint(values))
        ):
            raise ValueError("Kang counts must be finite nonnegative integers")
        obs = source.obs.iloc[in_chunk]
        for local_index, (source_index, row) in enumerate(obs.iterrows()):
            cell = str(source_index)
            donor, condition = split_design[cell].split("\x1f")
            if str(row["ind"]) != donor:
                raise ValueError(f"Kang split donor mismatch for {cell}")
            if not _condition_matches(str(row["stim"]), condition):
                raise ValueError(f"Kang split condition mismatch for {cell}")
            state = str(row["cell"]).strip()
            if not state or state.lower() in {"na", "nan", "none"}:
                raise ValueError(f"Kang cell-state identity is missing for {cell}")
            key = donor + "\x1f" + condition
            vector = block[local_index, :]
            summed = (
                np.asarray(vector.sum(axis=0)).reshape(-1)
                if sparse.issparse(vector)
                else np.asarray(vector).reshape(-1)
            )
            counts[sample_index[key], :] += np.rint(summed).astype(np.int64)
            sample_cell_counts[key] += 1
            cell_rows.append(
                {
                    "cell_id": cell,
                    "sample_id": f"{donor}::{condition}",
                    "donor": donor,
                    "condition": condition,
                    "state": state,
                }
            )

    if len(cell_rows) != len(selection):
        raise ValueError(f"Kang {split} materialization lost cells")
    sample_rows = []
    for key in sample_keys:
        donor, condition = key.split("\x1f")
        n_cells = sample_cell_counts[key]
        if n_cells < 20:
            raise ValueError(f"Kang {split} sample {donor}/{condition} has fewer than 20 cells")
        sample_rows.append(
            {
                "sample_id": f"{donor}::{condition}",
                "donor": donor,
                "condition": condition,
                "n_cells": n_cells,
            }
        )

    output_dir.mkdir(parents=True, exist_ok=True)
    counts_path = output_dir / f"{split}_pseudobulk_counts.tsv"
    samples_path = output_dir / f"{split}_sample_metadata.tsv"
    cells_path = output_dir / f"{split}_cell_metadata.tsv"
    genes = [str(value) for value in source.var_names]
    with counts_path.open("w", encoding="utf-8", newline="") as handle:
        handle.write("gene\t" + "\t".join(row["sample_id"] for row in sample_rows) + "\n")
        for gene_index, gene in enumerate(genes):
            handle.write(
                gene
                + "\t"
                + "\t".join(str(int(value)) for value in counts[:, gene_index])
                + "\n"
            )
    _write_tsv(samples_path, ["sample_id", "donor", "condition", "n_cells"], sample_rows)
    _write_tsv(cells_path, ["cell_id", "sample_id", "donor", "condition", "state"], cell_rows)
    return {
        "counts_path": counts_path,
        "samples_path": samples_path,
        "cells_path": cells_path,
        "n_cells": len(cell_rows),
        "n_genes": len(genes),
        "n_samples": len(sample_rows),
        "n_donors": len({row["donor"] for row in sample_rows}),
    }


def _validate_outputs(output_dir: Path) -> dict[str, int]:
    required = {
        "edger_reference.tsv": {"gene", "logFC", "F", "PValue", "FDR"},
        "edger_design_matrix.tsv": {"sample_id"},
        "propeller_reference.tsv": {
            "cluster", "baseline_proportion", "target_proportion", "effect", "PValue", "FDR"
        },
        "propeller_design_matrix.tsv": {"sample_id"},
        "method_null_reference.tsv": {"permutation_id", "gene", "PValue", "FDR"},
        "replicate_null_reference.tsv": {
            "permutation_id", "type1_rate", "null_effect_bias", "exchangeability_violation_count"
        },
    }
    counts: dict[str, int] = {}
    for name, columns in required.items():
        fields, rows = _read_tsv(output_dir / name)
        missing = sorted(columns - set(fields))
        if missing:
            raise ValueError(f"{name} is missing: {missing}")
        if not rows:
            raise ValueError(f"{name} is empty")
        counts[name] = len(rows)
        for row in rows:
            for column in columns & {"PValue", "FDR", "type1_rate"}:
                value = float(row[column])
                if not math.isfinite(value) or not 0 <= value <= 1:
                    raise ValueError(f"{name} has invalid {column}")
    _, replicate_rows = _read_tsv(output_dir / "replicate_null_reference.tsv")
    if any(int(row["exchangeability_violation_count"]) != 0 for row in replicate_rows):
        raise ValueError("REF-05 replicate null violated donor exchangeability")
    return counts


def generate(
    datasets_path: Path,
    splits_path: Path,
    repo: Path,
    rscript: Path,
    output_dir: Path,
) -> dict[str, Any]:
    import anndata as ad

    datasets = json.loads(datasets_path.read_text(encoding="utf-8"))
    record = datasets["datasets"][DATASET_ID]
    artifact = Path(record["artifact_path"]).resolve()
    if not artifact.is_file() or _sha256(artifact) != record["artifact_sha256"]:
        raise ValueError("Kang artifact is missing or has drifted")
    calibration = _selection_rows(splits_path, "calibration")
    evaluation = _selection_rows(splits_path, "evaluation")
    calibration_ids = {row["cell_id"] for row in calibration}
    evaluation_ids = {row["cell_id"] for row in evaluation}
    if calibration_ids & evaluation_ids:
        raise ValueError("Kang calibration/evaluation cell overlap")
    calibration_donors = {row["unit_id"] for row in calibration}
    evaluation_donors = {row["unit_id"] for row in evaluation}
    if calibration_donors & evaluation_donors:
        raise ValueError("Kang calibration/evaluation donor overlap")

    output_dir.mkdir(parents=True, exist_ok=True)
    inputs_dir = output_dir / "reference_inputs"
    source = ad.read_h5ad(artifact, backed="r")
    try:
        evaluation_inputs = _materialize_split_inputs(
            source, evaluation, inputs_dir, "evaluation"
        )
        calibration_inputs = _materialize_split_inputs(
            source, calibration, inputs_dir, "calibration"
        )
    finally:
        if getattr(source, "file", None) is not None:
            source.file.close()

    runner = (repo / "scripts" / "generate_paper_ref05.R").resolve()
    if not runner.is_file() or not rscript.is_file():
        raise FileNotFoundError(runner if not runner.is_file() else rscript)
    config = {
        "schema_version": "pertura-paper-ref05-r-config-v1",
        "baseline": BASELINE,
        "target": TARGET,
        "seed": SEED,
        "evaluation_counts": str(evaluation_inputs["counts_path"]),
        "evaluation_samples": str(evaluation_inputs["samples_path"]),
        "evaluation_cells": str(evaluation_inputs["cells_path"]),
        "calibration_counts": str(calibration_inputs["counts_path"]),
        "calibration_samples": str(calibration_inputs["samples_path"]),
        "output_dir": str(output_dir.resolve()),
    }
    config_path = inputs_dir / "ref05-r-config.json"
    _write_json(config_path, config)
    print("REF-05-A/B/C: running independent donor-aware R references", flush=True)
    completed = subprocess.run(
        [str(rscript), "--vanilla", str(runner), str(config_path)],
        cwd=repo,
        timeout=3600,
        check=False,
    )
    if completed.returncode != 0:
        raise RuntimeError(
            f"REF-05 R reference failed with exit code {completed.returncode}; "
            "inspect the streamed benchmark log"
        )
    row_counts = _validate_outputs(output_dir)
    environment_path = output_dir / "r_environment.json"
    environment = json.loads(environment_path.read_text(encoding="utf-8"))

    output_names = [
        "edger_reference.tsv",
        "edger_design_matrix.tsv",
        "edger_session_info.txt",
        "propeller_reference.tsv",
        "propeller_design_matrix.tsv",
        "propeller_session_info.txt",
        "method_null_reference.tsv",
        "replicate_null_reference.tsv",
        "r_environment.json",
    ]
    manifest = {
        "schema_version": SCHEMA_VERSION,
        "reference_pack_id": "REF-05",
        "readiness": "generated",
        "completed_jobs": ["REF-05-A", "REF-05-B", "REF-05-C"],
        "pending_jobs": [],
        "independent_of_pertura_results": True,
        "input_files": {
            "datasets.json": _sha256(datasets_path),
            "splits.json": _sha256(splits_path),
            "kang_artifact": _sha256(artifact),
            "r_runner": _sha256(runner),
            **{
                f"evaluation_{name}": _sha256(path)
                for name, path in (
                    ("counts", evaluation_inputs["counts_path"]),
                    ("samples", evaluation_inputs["samples_path"]),
                    ("cells", evaluation_inputs["cells_path"]),
                )
            },
            **{
                f"calibration_{name}": _sha256(path)
                for name, path in (
                    ("counts", calibration_inputs["counts_path"]),
                    ("samples", calibration_inputs["samples_path"]),
                )
            },
        },
        "output_files": {name: _sha256(output_dir / name) for name in output_names},
        "counts": {
            "evaluation_cells": evaluation_inputs["n_cells"],
            "evaluation_donors": evaluation_inputs["n_donors"],
            "calibration_cells": calibration_inputs["n_cells"],
            "calibration_donors": calibration_inputs["n_donors"],
            "genes": evaluation_inputs["n_genes"],
            "edger_rows": row_counts["edger_reference.tsv"],
            "propeller_rows": row_counts["propeller_reference.tsv"],
            "method_null_rows": row_counts["method_null_reference.tsv"],
            "replicate_null_rows": row_counts["replicate_null_reference.tsv"],
            "exchangeability_violation_count": 0,
            "cell_label_permutation_count": 0,
        },
        "parameters": {
            "seed": SEED,
            "baseline": BASELINE,
            "target": TARGET,
            "analysis_unit": "donor",
            "design": "~ donor + condition",
            "null_permutation_unit": "paired_donor_condition_label",
            "cell_label_permutation": False,
        },
        "environment": environment,
        "limitations": [
            "Kang is replicated stimulation scRNA-seq and is not presented as Perturb-seq.",
            "edgeR and Propeller references use four held-out evaluation donors.",
            "Null references use mixed within-donor condition swaps on calibration donors; cells are never permuted.",
            "REF-05 distinguishes composition changes from within-state expression effects.",
        ],
    }
    manifest_path = output_dir / "manifest.json"
    _write_json(manifest_path, manifest)
    print("REF-05: manifest written", flush=True)
    return {
        "schema_version": SCHEMA_VERSION,
        "reference_pack_id": "REF-05",
        "readiness": "generated",
        "completed_jobs": manifest["completed_jobs"],
        "pending_jobs": [],
        "passed": True,
        "problems": [],
        "manifest_sha256": _sha256(manifest_path),
    }


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Generate independent REF-05 edgeR, Propeller, and null references."
    )
    parser.add_argument("--datasets", type=Path, required=True)
    parser.add_argument("--splits", type=Path, required=True)
    parser.add_argument("--repo", type=Path, required=True)
    parser.add_argument("--rscript", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    result = generate(
        args.datasets.resolve(),
        args.splits.resolve(),
        args.repo.resolve(),
        args.rscript.resolve(),
        args.output.resolve(),
    )
    print(json.dumps(result, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
